SecondBridge/SecondBridge/Data/BinarySearchTree.swift

```diff
@@ -0,0 +1,236 @@
+/*
+* Copyright (C) 2015 47 Degrees, LLC http://47deg.com hello@47deg.com
+*
+* Licensed under the Apache License, Version 2.0 (the "License"); you may
+* not use this file except in compliance with the License. You may obtain
+* a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+import Foundation
+
+internal enum PartsTree {
+    case Left
+    case Right
+}
+
+indirect enum BinarySearchTree<T:Comparable>: CustomStringConvertible {
+    
+    case Empty
+    case Node(T, left: BinarySearchTree, right: BinarySearchTree)
+    
+    
+    /*
+    * Return the element the head the current BST
+    */
+    func getHead() -> T? {
+        switch self{
+        case .Empty:
+            return .None
+        case let .Node(T, left: _, right: _):
+            return T
+        }
+    }
+    
+    /*
+    * Return a new BST with the part provided in the parameter (left or right)
+    */
+    func get(part: PartsTree) -> BinarySearchTree {
+        
+        switch self {
+        case .Empty:
+            return .Empty
+        case let .Node(_, left: _, right: rigth) where part == .Right:
+            return rigth
+        case let .Node(_, left: left, right: _) where part == .Left:
+            return left
+        default:
+            return self
+        }
+    }
+    
+    /**
+    * Returns a new BinarySearchTree with the current contents and the provided item.
```

Comment:
Returns a new BinarySearchTree with the provided item added to the current contents.